### ISSUE TITLE
Save local timezone

### ISSUE THREAD

**Issue by gwarnants** [mendhak/gpslogger#911]
Hello,

As correctly explained [here](https://gpslogger.app/#itsusingthewrongtimezonecanyoumakeitusemytimezone), when saving to GPX, KML, or custom URL, points are logged using UTC time.

If we want to use the log file in any application, it means we must remember ourself where the log was recorded (I mean in which timezone). If I recorded an activity at the other end of the world, I'll probably want to display that activity using the timezone of the country I was, not the timezone of the country I'm currently in.

These files saved by GPSLogger currently do not give any way to know the local timezone. If I forget in which country the log was recorded, there's no way convert logs to local dates; except using a third-party webservice ([like theses](https://stackoverflow.com/questions/16086962/how-to-get-a-time-zone-from-a-location-using-latitude-and-longitude-coordinates)), which are often not free to use.

I think it would be nice to add an extra information to allow these date conversions

Thanks

**Comment by dshanske** [comment]
I'd like this as an option in the custom URL

**Comment by mendhak** [comment]
I'm trying to think of the least friction way of doing this. I'm aware GPX extensions exist, but there's no guarantee that all applications read what I come up with, or if there are any well known GPX extensions for storing an additional piece of info like this. I wasn't able to find anything well known to store offsets. 

Would it be useful instead if I introduced some toggle like "Record local timezone anyway".  If you toggle it, then the `<time>` will be written with the offset, instead of the UTC `Z`.  In other words, I drop this strictness highlighted in the FAQ item. The user can knowingly choose to have the timezone offset in the GPX file even if it's 'incorrect'.  I say 'incorrect' because after some searching I'm not sure if this is enforced or not.  


Some validation will be required right now, what you could do is test your existing GPX files, edit them to have an offset, and see if your applications accept it. 

Like: 

```
2021-09-26T11:55:36.7+01:00
2021-09-26T11:55:36.700+01:00
```


At a glance there seem to be some classes in Java that can help me here, but it will require Android 24+ or Android 26+ (Android 7+ or 8+).   For my own reference the ZonedDateTime could help me greatly. I can take the location epoch passed from the OS, and convert it to the user's time.  Then easily write it out to the formats shown above. 

```
if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
    ZonedDateTime zoned = ZonedDateTime.ofInstant(Instant.ofEpochMilli(loc.getTime()), TimeZone.getDefault().toZoneId());
    String withOffset1 = zoned.toOffsetDateTime().toString();
    String withOffset2 = zoned.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
}
```

**Comment by MexasPT** [comment]
I've had the same problem for a month...
I'm importing csv into a mysql table, but I wanted to add 1 hour... 
Someone explain to me how to add a time in > $data[0] ?!

$data[0] is the coll 2021-09-27T10:46:09

Thanks!!!

CODE:
`$speedcorr = 3.6 * $data[6];

    $query = 'INSERT INTO positions (time, v_id, latitude, longitude, altitude, speed, bearing, accuracy, provider, comment, created_date) 
`VALUES ("'.**$data[0]** . '", "1", "' . $data[1].'", "' . $data[2] . '", "' . $data[3] . '", "' . $speedcorr . '", "' . $data[5] . '", "' . $data[4] . '", "' . $data[8] . '", "' . $data[16] . '", "' . $data[0] . '")';`

**Comment by mendhak** [comment]
@MexasPT I think that's PHP? 

I am not PHP savvy but something like this should work.  DateTime should be able to read ISO8601 strings.

```
$dt = new DateTime($data[0]);
$dt->modify('+ 1 hour');
$result = $dt->format(DateTime::ATOM);

$query = 'INSERT INTO positions (time, v_id, latitude, longitude, altitude, speed, bearing, accuracy, provider, comment, created_date) 
VALUES ("'.$result. '", "1", "' . $data[1].'", "' . $data[2] . '", "' . $data[3] . '", "' . $speedcorr . '", "' . $data[5] . '", "' . $data[4] . '", "' . $data[8] . '", "' . $data[16] . '", "' . $data[0] . '")';
```

**Comment by mendhak** [comment]
I might be able to use this, should work on older Android versions too.  The Calendar is used just to figure out the timezone. 

```
    public static String getIsoDateTimeWithOffset(Date dateToFormat){
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        SimpleDateFormat ISO_8601_DATE_TIME = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ", Locale.US);
        ISO_8601_DATE_TIME.setTimeZone(calendar.getTimeZone());
        String dateTimeStringWithOffset = ISO_8601_DATE_TIME.format(dateToFormat);
        //Fix for older Android where ZZZZZ does not include colon
        if (!dateTimeStringWithOffset.substring(dateTimeStringWithOffset.length() - 3).startsWith(":")) {
            dateTimeStringWithOffset = dateTimeStringWithOffset.substring(0, dateTimeStringWithOffset.length() - 2) + ":" + dateTimeStringWithOffset.substring(dateTimeStringWithOffset.length() - 2);
        }
        return dateTimeStringWithOffset;
    }

```

**Comment by mendhak** [comment]
Today I played a bit with a GPX file, I deliberately modified all the timestamps to have a timezone offset in it.  I then tested it with exiftool, [GPXSee](https://www.gpxsee.org/), [GPSVisualizer](https://www.gpsvisualizer.com/), [UTrack](http://utrack.crempa.net/).  

They were all able to read the times with the offsets.  So I'm thinking to add a setting to just go ahead and give the option.  In the 'logging preferences' screen, how's something like this (default false)

![image](https://user-images.githubusercontent.com/746276/135930076-af8d9c61-af72-4aa2-9dc5-277a22ecdbd2.png)
